### PR TITLE
feat(Keystore): Introduce additional KMSConfiguration options

### PR DESCRIPTION
**Pull request by texastony**
_Issue #, if available:_

_Description of changes:_
The Hierarchical Keyring's Keystore now supports four (4) `KMSConfigurations`:
- kmsKeyArn
- kmsMRKeyArn
- discovery
- MRDiscovery

See our Examples and JavaDocs for details
on how these options effect the relationship between a Keystore and KMS.

_Squash/merge commit message, if applicable:_

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+  const NO_CORRESPONDING_BRANCH_KEY :=
+    "No item found for corresponding branch key identifier."
+
+  const BRANCH_KEY_ID_NEEDED :=
+    "Empty string not supported for branch key identifier."
+
+  // If the Item/Record contains an invalid KMS ARN
+  const RETRIEVED_KEYSTORE_ITEM_INVALID_KMS_ARN :=
+    "The `kms-arn` field for the requested Branch Key ID is corrupted."
```
Comment (justplaz) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy:38]: nit: consistency - should this be "branch key identifier" to match other messages?

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy [hunk 2]
```diff
@@ -0,0 +1,39 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+module {:options "/functionSyntax:4" } KeyStoreErrorMessages {
+  const KMS_CONFIG_KMS_ARN_INVALID :=
+    "KMSConfiguration's KMS Key ARN is invalid."
+
+  const CUSTOM_BRANCH_KEY_ID_NEED_EC :=
+    "Custom branch key id requires custom encryption context."
```
Comment (justplaz) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy:9]: nit: consistency - should this be "branch key identifier" to match other messages?

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  const GET_KEY_ARN_DISAGREEMENT :=
+    "Branch Key's AWS KMS Key ARN read from Dynamodb does not match Key Store's configured AWS KMS Key ARN."
```
Comment (justplaz) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy:12]: nit: consistency - we don't use AWS KMS, just KMS elsewhere

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/test/TestVersionKey.dfy [hunk 4]
```diff
@@ -88,6 +90,105 @@ module TestVersionKey {
...
+    expect mat.branchKeyIdentifier == id;
+    var mat3EC :- expect DecodeEncryptionContext(mat.encryptionContext);
+    expect mat3EC == customEC;
+    expect mat3.branchKeyVersion != mat.branchKeyVersion;
+
+    var newActiveVersion :- expect UTF8.Decode(newActiveResult.branchKeyMaterials.branchKeyVersion);
+
+    expect newActiveVersion != oldActiveVersion;
+
+    // Since this process uses a read DDB table,
```
Comment (justplaz) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/test/TestVersionKey.dfy:173]: nit: s/read/real

**Review by justplaz — COMMENTED** [review]
Overall comment: Are we adding examples for this?

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy (hunk 1 above)
Comment (texastony) [reply AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy:38]: But it's not the Branch Key Identifier that is malformed.
It is the `kms-arn` field.

**Comment by texastony** [comment]
> Overall comment: Are we adding examples for this?

There is no example section of this repo.
I am going to add examples to the DB-ESDK and ESDK.

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy (hunk 1 above)
Comment (texastony) [reply AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy:38]: OH, yeah, you are right

Comment (texastony) [reply AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy]: Resolved in 6bac41f3

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy (hunk 2 above)
Comment (texastony) [reply AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy]: Resolved in 6bac41f3

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy (hunk 3 above)
Comment (texastony) [reply AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/ErrorMessages.dfy]: Resolved in 6bac41f3

**Review by justplaz — DISMISSED** [review]
LGTM!

**Review by justplaz — APPROVED** [review]
re-approving

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/Model/KeyStore.smithy [hunk 5]
```diff
@@ -95,18 +95,44 @@ structure KeyStoreConfig {
   id: String,
   @javadoc("The AWS KMS grant tokens that are used when this Key Store calls to AWS KMS.")
   grantTokens: GrantTokenList,
-  @javadoc("The DynamoDB client this Key Store uses to call Amazon DynamoDB. If None is provided, a default DynamoDB Client is created using the Region from the KMS ARN.")
+  @javadoc("The DynamoDB client this Key Store uses to call Amazon DynamoDB. If None is provided and the KMS ARN is, the KMS ARN is used to determine the Region of the default client.")
```
Comment (seebees) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/Model/KeyStore.smithy:98]: `and the KMS ARN is, the ` I think that needs to be updated.

Comment (seebees) [reply AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/Model/KeyStore.smithy:98]: No, I get it now. Fine :)

**Pull request merged**